rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,441 @@
+as big as possible to achieve best performance. The default workspace size is
+set to device global memory size from v8.4 release.
+
+## Proposal
+
+### Dedicated cache for different engine kind
+In oneDNN Graph programming model, different graph can have different engine
+kind and simultaneously run on different engine, like cpu or gpu. Different kind
+of engine usually has different memory limitation. 
+
+So, we propose to use dedicated cache for different engine kind, and let users
+set different capacity for different engine kind separately.
+
+For multiple devices scenarios, like multi-GPU, whether to use dedicated cache
+for each device is still an open question, we will discuss it in later sections.
+
+### Unified cache for all backends
+We propose to implement an unified constant tensor cache in the oneDNN Graph
+interface layer. Each backend should use this interface to cache their all
+constant tensors into the unified cache. And the oneDNN Graph API users could
+control the capacity of the unified cache, no need to be aware of any backend
+specific things. This proposal is based on the following considerations:
+
+- In memory limited scenarios, we usually want to control the total memory
+  footprint, instead of the memory footprint of specific backend.
+- By the design of oneDNN Graph API, the API users don't need to be aware about
+  which backend is using. We don't like to break this design currently.
+- Backend selection is dynamic in mixed backend scenarios. Users can't
+  explicitly know which backend will run and how many partition will a backend
+  claims until graph partition stage done. So, it's hard to decide how many
+  memory a backend should use.
+
+### Capacity setting/getting frontend API
+Based on above considerations, we proposed to add the following two C++ API and
+corresponding C API to control the capacity of constant tensor cache:
+
+```cpp
+void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, size_t capacity);
+size_t get_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind);
+```
+
+The unit of capacity is `MBytes`. Users can call the setter API to set the
+capacity for specific engine kind. If the cached tensors' size reaches the
+capacity, the new coming tensors won't be cached while the cached tensor won't
+be evicted by default. The getter API can be used to query the current capacity
+of specific engine kind. Please note that the specified capacity is just a hint,
+and the library does not guarantee that the cached tensors size will always
+reach the capacity.
+
+The default capacity is kept to be infinite on both CPU and GPU to keep the
+library default behavior backward compatible.
+
+We also allow to control the capacity by using an environment variable named
+`ONEDNN_GRAPH_CONSTANT_TENSOR_CACHE_CAPACITY`. The form of this env var value is
+`engine_kind:size`. We can use `;` to split the configuration for different
+engine kind. Below is an example:
```

Comment:
In addition to programmable API there is an environment variable named
`ONEDNN_GRAPH_CONSTANT_TENSOR_CACHE_CAPACITY` to control the capacity. It accepts values in form `engine_kind:size` or `engine_kind1:size1;engine_kind2:size2`. Example: